Add block sparse linear and locally connected layers

clane9 opened the pull request.

clane9 committed:
Initial blocksparse linear outline

alismil committed:
BlockSparseLinear complete

alismil committed:
added MixtureLayerNorm

alismil committed:
matched MixtureLinear to main

alismil committed:
completed BlockSparseLocallyConnected

alismil committed:
updated comments in _create_connectivity_matrix

clane9 committed:
Update blocksparse linear implementation

- Use pytorch native blocksparse tensors following [this
  blog](https://pytorch.org/blog/speeding-up-vits/) rather than triton
  blocksparse, which seems not very stable. See e.g. [this recent
  PR](https://github.com/triton-lang/triton/pull/4156) where
  `triton.ops` were deprecated.
- Represent weights in `BlockSparseLinear` as block-sparse tensor rather
  than dense tensor. This will save a lot of gpu memory.
- Change `BlockSparseLocallyConnected` interface to more closely match
  `nn.Conv2d`. Except remove support for padding and stride. For now we
  should be able to restrict to `padding="same"`, `stride=1` (?)
- Rewrite function to construct local connectivity matrix to directly
  construct a sparse rather than dense matrix. Use vectorized
  operations rather than for loops for construction. This should save a
  lot of memory and run faster.
- Add support for multiple input and output channels and depthwise
  convolution. The channels axis can either be first or last. For
  depthwise convolution, first should be more efficient (more block
  sparsity).

TODO:
- Finish testing on cuda. native blocksparse matmul is not implemented on CPU.

clane9 commented:
I updated the implementation:

- Use pytorch native blocksparse tensors following [this blog](https://pytorch.org/blog/speeding-up-vits/) rather than triton blocksparse, which seems not yet stable. 
- Represent weights in `BlockSparseLinear` as block-sparse tensor rather than dense tensor. This will save a lot of gpu memory.
- Change `BlockSparseLocallyConnected` interface to more closely match `nn.Conv2d`. Except remove support for padding and stride. For now we should be able to restrict to `padding="same"`, `stride=1` (@alismil what do you think?).
- Rewrite function for constructing local connectivity matrix to directly construct a sparse rather than dense matrix. Use vectorized operations rather than for loops for construction. This should save a lot of memory and run faster.
- Add support for multiple input and output channels and depthwise convolution. The channels axis can either be first or last. For depthwise convolution, first should be more efficient (more block sparsity).

alismil commented on columnformers/models/layers.py:
```diff
@@ -280,3 +283,232 @@ def init_weights(module: nn.Module):
+            input, f"{in_pattern} -> {out_pattern}", h=self.height, w=self.height
+        )
+        output = self.bsl(output)
+        output = rearrange(
```
Does it make sense to keep rearranging the input on each forward pass? Might be slightly more efficient if we assume all inputs are batches of flattened images and then have a separate function to rearrange when necessary (e.g. to visualise activations etc)?

clane9 replied:
Yes good point. If you have `in_shape="nlc"` and `channels_last=True` (or `in_shape="nchw"` and `channels_last=False`) the rearrange will just produce a new view without actually moving the data. Otherwise however, it will move the data a lot like you say.

My thinking was that the rearranging helps the `BlockSparseLocallyConnected` be consistent with `nn.Conv2d` (when `in_shape="nchw"`) and with `nn.Linear` (when `in_shape="nlc"`).

What if we add a third `in_shape` option `nd` that assumes the input is already flattened?

alismil replied:
Yes I think that's a good idea?

alismil commented:
> For now we should be able to restrict to `padding="same"`, `stride=1` (@alismil what do you think?).

If we want to use this for All TNNs, to make the dimensions work out we have to use different strides for each layer

clane9 committed:
Debug block sparse linear

Could not use `sparse_bsr` layout weight parameter. It fails to map to
cuda correctly when calling `model.cuda()`. This is probably a bug that
should be reported. But as a workaround I just unpack the
`crow_indices`, `col_indices` as buffers and store the sparse bsr weight
values as a standard strided parameter. Then I construct the sparse bsr
weight tensor on the fly during forward.

TODO: backward does not work. Raises

```
RuntimeError: addmm: computation on CUDA is not implemented for Strided + Strided @ SparseBsr
```

clane9 commented:
> If we want to use this for All TNNs, to make the dimensions work out we have to use different strides for each layer

Ok good point. Then maybe we should add back the support you had for strides and padding to `_sparse_local_connectivity`. Hopefully it doesn't complicate things too much.

clane9 committed:
Add `in_shape="nd"` option for no rearranging

alismil commented on columnformers/models/layers.py:
```diff
@@ -280,3 +283,240 @@ def init_weights(module: nn.Module):
+
+    # conv kernel index offsets. note that the kernel width is required to be odd.
+    # (k^2, 2)
+    kernel_half_width = (kernel_size - 1) // 2
```
This is a bit restrictive, can we adapt this to also include even kernel size by doing something like this?

```
if kernel_size % 2 == 0:
    kernel_half_width = kernel_size // 2
    kernel_indices = torch.cartesian_prod(
        torch.arange(-kernel_half_width, kernel_half_width),
        torch.arange(-kernel_half_width, kernel_half_width),
    )
else:
    kernel_half_width = (kernel_size - 1) // 2
    kernel_indices = torch.cartesian_prod(
        torch.arange(-kernel_half_width, kernel_half_width + 1),
        torch.arange(-kernel_half_width, kernel_half_width + 1),
    )
```

clane9 replied:
Yes I think you're right, something like this would probably be better. Although it feels like it should be possible to make the code shorter.

More generally, it would probably be best to have exactly the same interface and behavior as [Conv2d](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html). What I have now takes a few shortcuts.
